testng-eclipse-plugin/src/main/org/testng/eclipse/launch/TestFinder.java

```diff
@@ -0,0 +1,155 @@
+package org.testng.eclipse.launch;
+
+import java.util.HashSet;
+import java.util.Set;
+
+import org.eclipse.core.runtime.CoreException;
+import org.eclipse.core.runtime.IProgressMonitor;
+import org.eclipse.core.runtime.NullProgressMonitor;
+import org.eclipse.core.runtime.SubProgressMonitor;
+import org.eclipse.jdt.core.Flags;
+import org.eclipse.jdt.core.IJavaElement;
+import org.eclipse.jdt.core.IRegion;
+import org.eclipse.jdt.core.ISourceRange;
+import org.eclipse.jdt.core.IType;
+import org.eclipse.jdt.core.ITypeHierarchy;
+import org.eclipse.jdt.core.JavaCore;
+import org.eclipse.jdt.core.JavaModelException;
+import org.eclipse.jdt.core.dom.AST;
+import org.eclipse.jdt.core.dom.ASTNode;
+import org.eclipse.jdt.core.dom.ASTParser;
+import org.eclipse.jdt.core.dom.CompilationUnit;
+import org.eclipse.jdt.core.dom.IAnnotationBinding;
+import org.eclipse.jdt.core.dom.IBinding;
+import org.eclipse.jdt.core.dom.IMethodBinding;
+import org.eclipse.jdt.core.dom.ITypeBinding;
+import org.eclipse.jdt.core.dom.Modifier;
+import org.eclipse.jdt.core.dom.TypeDeclaration;
+import org.eclipse.jdt.core.search.IJavaSearchConstants;
+import org.eclipse.jdt.core.search.IJavaSearchScope;
+import org.eclipse.jdt.core.search.SearchEngine;
+import org.eclipse.jdt.core.search.SearchParticipant;
+import org.eclipse.jdt.core.search.SearchPattern;
+import org.eclipse.jdt.core.search.SearchRequestor;
+import org.eclipse.jdt.internal.junit.util.CoreTestSearchEngine;
+import org.testng.eclipse.util.ResourceUtil;
+
+public class TestFinder {
+
+  public void findTestsInContainer(IJavaElement element, Set result, IProgressMonitor pm) throws CoreException {
+    if (element == null || result == null) {
+      throw new IllegalArgumentException();
+    }
+
+    if (element instanceof IType) {
+      if (internalIsTest((IType) element, pm)) {
+        result.add(element);
+        return;
+      }
+    }
+
+    if (pm == null)
+      pm= new NullProgressMonitor();
+
+    try {
+      pm.beginTask(ResourceUtil.getString("TestSearchEngine.message.searching"), 4);
+
+      IRegion region= CoreTestSearchEngine.getRegion(element);
+      ITypeHierarchy hierarchy= JavaCore.newTypeHierarchy(region, null, new SubProgressMonitor(pm, 1));
+      IType[] allClasses= hierarchy.getAllClasses();
+
+      // search for all types with references to RunWith and Test and all subclasses
+      HashSet<IType> candidates= new HashSet<>(allClasses.length);
+      SearchRequestor requestor= new AnnotationSearchRequestor(hierarchy, candidates);
+
+      IJavaSearchScope scope= SearchEngine.createJavaSearchScope(allClasses, IJavaSearchScope.SOURCES);
```

Comment:
the benefit of JDT search engine is we can easily specify the Search scope as SOURCES and we can benefit from the indexing of JDT search